File: week-2/problem-set/prob5.py
"""Prob5 example.

This file belongs to the week 2 data-structure and graph practice solutions. It uses is_anagram to illustrate the main idea with a small runnable example.
"""

def is_anagram(word1, word2):
    # Convert both words to lowercase and remove non-letter characters
    """Return True when the two words contain the same letters.

    Args:
        word1 (str): The word1 input used by this function.
        word2 (str): The word2 input used by this function.

    Returns:
        bool: True when the required condition is satisfied, otherwise False.
    """
    word1 = ''.join(c for c in word1.lower() if c.isalpha())
    word2 = ''.join(c for c in word2.lower() if c.isalpha())
    
    # Sort the letters of each word in alphabetical order
    sorted_word1 = sorted(word1)
    sorted_word2 = sorted(word2)
    
    # Compare the sorted letter sequences of the two words
    return sorted_word1 == sorted_word2

print(is_anagram("tea", "eat"))  # True
print(is_anagram("silent", "listen"))  # True
print(is_anagram("desserts", "stressed"))  # True
print(is_anagram("hello", "world"))  # False

File: week-2/problem-set/prob5.cpp
/**
 * @file prob5.cpp
 * @brief C++ version of the anagram example.
 */

#include <algorithm>
#include <cctype>
#include <iostream>
#include <string>

using namespace std;

/**
 * @brief Return true when two words contain the same letters.
 * @param word1 string First text to compare.
 * @param word2 string Second text to compare.
 * @return bool True when the cleaned and sorted letters match, otherwise false.
 */
bool is_anagram(string word1, string word2) {
    string cleaned1;
    string cleaned2;

    for (char character : word1) {
        if (isalpha(static_cast<unsigned char>(character))) {
            cleaned1.push_back(static_cast<char>(tolower(static_cast<unsigned char>(character))));
        }
    }
    for (char character : word2) {
        if (isalpha(static_cast<unsigned char>(character))) {
            cleaned2.push_back(static_cast<char>(tolower(static_cast<unsigned char>(character))));
        }
    }

    sort(cleaned1.begin(), cleaned1.end());
    sort(cleaned2.begin(), cleaned2.end());
    return cleaned1 == cleaned2;
}

/**
 * @brief Run the anagram demonstrations.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    cout << boolalpha << is_anagram("tea", "eat") << '\n';
    cout << is_anagram("silent", "listen") << '\n';
    cout << is_anagram("desserts", "stressed") << '\n';
    cout << is_anagram("hello", "world") << '\n';
    return 0;
}
